### PR TITLE
Add ability to list key version destruction jobs, recurring jobs and destroying key versions

### PR DESCRIPTION
**Pull request by johanbrandhorst**
## [feat(kms): Add ability to list data key version destruction jobs](https://github.com/hashicorp/boundary/commit/b1f2e2f2683cb0891489934e7c53966df47e2fe4)

Adds the domain part of this new method

## [feat(scopes): Add ListKeyVersionDestructionJobs](https://github.com/hashicorp/boundary/commit/5c8e5f2a7d08014f1724058562ca64aa3afb2602)

Adds the API part of this new method

## [feat(cli): Add list-key-version-destruction-jobs](https://github.com/hashicorp/boundary/commit/7f4ffa69069811b67f9503416ea8d8bfcdf8a782)

Adds the CLI part of this new method

## [feat(kms): Add recurring job that performs table rewrapping](https://github.com/hashicorp/boundary/commit/c4f7fb875b0201276c9de6f313b3bd43470a6ee9)

The new recurring job runs for each job table with a registered
rewrapping callback. It attempts to become the running run and
start its rewrapping, gracefully handling sudden interruptions
by resuming it's work if it finds evidence of sudden interruption.

## [feat(kms): Add recurring job for destroying key versions](https://github.com/hashicorp/boundary/commit/9d8eada4bdd05c0091eb32399a727757d659b853)

The new job monitors the database for data key version destruction
jobs that have finished rewrapping all its data, performing
the final data key version revocation.

## [feat(kms): Add DestroyKeyVersion](https://github.com/hashicorp/boundary/pull/2498/commits/31f0e8d576d83c4887fe7f2dcb39842fde3a0dd0)

DestroyKeyVersion immediately destroys a key version
if it is a root key version or a data key version
which encrypts no data. If the data key version
encrypts data which needs to be rewrapped, it
queues an asynchronous job to complete the rewrapping
operation.

## [feat(cli): Add scopes destroy-key-version CLI command](https://github.com/hashicorp/boundary/pull/2498/commits/ce458b2c27f9b68a933c2f3d557b8c3a798878fc)

### CODE REVIEW

File: internal/cmd/commands/scopescmd/list_key_version_destruction_jobs.go [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+			output = append(output,
+				fmt.Sprintf("    Rows completed:  %d", item.CompletedCount),
+			)
+		}
+		if true {
+			output = append(output,
+				fmt.Sprintf("    Total rows:      %d", item.TotalCount),
+			)
+		}
+	}
```
Comment (johanbrandhorst) [inline internal/cmd/commands/scopescmd/list_key_version_destruction_jobs.go:102]: This is a proposal for how to display these on the CLI. This is what it looks like on the CLI:

```
$ boundary scopes list-key-version-destruction-jobs -scope-id global

Key version destruction job information:
  Key version ID:    kdkv_OJz1sJIgbZ
    Scope ID:        global
    Started:         Mon, 26 Sep 2022 17:23:07 PDT
    Status:          pending
    Rows completed:  0
    Total rows:      300
```

File: internal/api/genapi/input.go [hunk 2]
```diff
@@ -167,6 +167,18 @@ var inputStructs = []*structInfo{
...
+		inProto:     &scopes.KeyVersionDestructionJob{},
+		outFile:     "scopes/key_version_destruction_job.gen.go",
+		skipOptions: true,
+		fieldOverrides: []fieldInfo{
+			// int64 fields get marshalled by protobuf as strings, so we have
+			// to tell the json parser that their json representation is a
+			// string but they go into Go int64 types.
+			{Name: "CompletedCount", JsonTags: []string{"string"}},
+			{Name: "TotalCount", JsonTags: []string{"string"}},
+		},
```
Comment (johanbrandhorst) [inline internal/api/genapi/input.go:180]: Note that this is taking advantage of a feature of the CLI generation that will likely change soon - it was merged for 0.11.0 to get support for stringified large integers into 0.11.0 but longer term this will likely be implicit behavior when using large integers in the protobuf type. CC @tmessi @hugoamvieira

File: internal/cmd/commands/scopescmd/list_key_version_destruction_jobs.go [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+	return wordwrap.WrapString("List all pending key version destruction jobs within a scope", base.TermWidth)
+}
+
+func (c *ListKeyVersionDestructionJobsCommand) Help() string {
+	return base.WrapForHelpText([]string{
+		"Usage: boundary scopes list-key-version-destruction-jobs [args]",
+		"",
+		"  List all pending key version destruction jobs within a scope. A key version",
+		"  destruction job asynchronously re-encrypts any existing data with the latest",
+		"  key version before destroying the key version. Example:",
```
Comment (johanbrandhorst) [inline internal/cmd/commands/scopescmd/list_key_version_destruction_jobs.go:35]: I don't know if this is the right place to explain this sort of thing, but I figured it couldn't hurt, right?

File: internal/daemon/controller/handlers/scopes/scope_service_test.go [hunk 4]
```diff
@@ -1950,3 +1956,226 @@ func TestRotateKeys(t *testing.T) {
...
+		_, err = sqldb.ExecContext(context.Background(), "insert into kms_data_key_version_destruction_job_run(key_id, table_name, total_count) values ($1, 'auth_token', 100)", kvToDestroy.Id)
+		require.NoError(t, err)
+		_, err = sqldb.ExecContext(context.Background(), "insert into kms_data_key_version_destruction_job_run(key_id, table_name, total_count) values ($1, 'auth_oidc_method', 200)", kvToDestroy.Id)
+		require.NoError(t, err)
+		_, err = sqldb.ExecContext(context.Background(), "update kms_data_key_version_destruction_job_run set completed_count=100 where key_id=$1 and table_name='auth_token'", kvToDestroy.Id)
+		require.NoError(t, err)
+		t.Cleanup(func() {
+			_, err = sqldb.ExecContext(context.Background(), "delete from kms_data_key_version_destruction_job where key_id=$1", kvToDestroy.Id)
+			require.NoError(t, err)
+		})
```
Comment (johanbrandhorst) [inline internal/daemon/controller/handlers/scopes/scope_service_test.go:2052]: It's a layer violation to have the scope service test reach right into the database like this, but we don't have much choice until the real DestroyKeyVersion is available.

File: internal/proto/controller/storage/kms/store/v1/data_key_version_destruction_job_progress.proto [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+  // @inject_tag: `gorm:"not_null"`
+  string status = 50;
+
+  // The number of rows that have been rewrapped
+  // @inject_tag: `gorm:"not_null"`
+  int64 completed_count = 60;
+
+  // The total number of rows that need rewrapping
+  // @inject_tag: `gorm:"not_null"`
+  int64 total_count = 70;
```
Comment (johanbrandhorst) [inline internal/proto/controller/storage/kms/store/v1/data_key_version_destruction_job_progress.proto:33]: The use of `bigint` in our database schema necessitates this large integer

File: internal/tests/cli/boundary/scopes.bats [hunk 6]
```diff
@@ -24,6 +24,12 @@ export NEW_PROJECT='test_project'
 	[ "$status" -eq 0 ]
 }
 
+@test "boundary/scopes: can list key version destruction jobs at global scope" {
+	run list_key_version_destruction_jobs $DEFAULT_GLOBAL
+  echo "$output"
+	[ "$status" -eq 0 ]
+}
```
Comment (johanbrandhorst) [inline internal/tests/cli/boundary/scopes.bats:31]: These tests will become more interesting once we can request a key destruction

File: testing/internal/e2e/scopes/scope_test.go [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+	client, err := boundary.NewApiClient()
+	require.NoError(t, err)
+
+	ctx := context.Background()
+
+	scopeClient := scopes.NewClient(client)
+	keyVersionDestructionJobListResult, err := scopeClient.ListKeyVersionDestructionJobs(ctx, "global")
+	require.NoError(t, err)
+	assert.Empty(t, keyVersionDestructionJobListResult.Items)
+}
```
Comment (johanbrandhorst) [inline testing/internal/e2e/scopes/scope_test.go:49]: CC @moduli let me know if its premature to add tests to this framework.

Comment (moduli) [reply testing/internal/e2e/scopes/scope_test.go:49]: Nope! I think it should be ok. Two things...
1. We'll need to add an enos scenario for this. Based on what I see here, it looks like it will just be a duplicate of `enos-scenario-e2e-target.hcl`, but you can get rid of the `create_target` step and update `test_package` to point to this
2. There's a slight change to `RunCommand` coming in https://github.com/hashicorp/boundary/pull/2474 that will make it look something like this
```
output = e2e.RunCommand("boundary", e2e.WithArgs("scopes", "list-key-version-destruction-jobs",
		"-scope-id", "global",
		"-format", "json",
	))
```

Comment (moduli) [reply testing/internal/e2e/scopes/scope_test.go:49]: Also, maybe nit, but I might suggest changing the directory name from `scopes` to `scope`

File: testing/internal/e2e/scopes/scope_test.go [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+	"github.com/hashicorp/boundary/api/scopes"
+	"github.com/hashicorp/boundary/testing/internal/e2e"
+	"github.com/hashicorp/boundary/testing/internal/e2e/boundary"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+// TestListKeyVersionDestructionJobsCli uses the boundary CLI to list
+// pending key version destruction jobs.
+func TestListKeyVersionDestructionJobsCli(t *testing.T) {
```
Comment (tmessi) [inline testing/internal/e2e/scopes/scope_test.go:17]: Is there a workflow that can be tested here? Generally I think these tests shouldn't just test a single cli command, you already added tests for that in the cli (bats) tests.

Comment (johanbrandhorst) [reply testing/internal/e2e/scopes/scope_test.go:17]: There will be a workflow once we have the ability to trigger a key destruction. I could remove these tests until we can do them all together. I'm also not sure what the overlap between bats and these are, IMO we could replace the bats tests with this new framework.

Comment (tmessi) [reply testing/internal/e2e/scopes/scope_test.go:17]: My understanding of the differences between the tests is that the bats tests are for testing the cli ui, so testing individual commands and getting expected outputs, while theses tests are for testing end-to-end admin workflows that will often involve multiple commands, and could be accomplished via the cli/api/admin ui.

I agree that the cli ui tests *could* be implemented using some of this code, instead of using bats, but I don't think there are any current plans to do so.

File: testing/internal/e2e/scopes/scope_test.go (hunk 7 above)
Comment (johanbrandhorst) [reply testing/internal/e2e/scopes/scope_test.go]: I removed these for now, will readd once we have key destruction commands.

File: internal/daemon/controller/handlers/scopes/scope_service_test.go (hunk 4 above)
Comment (DanielleMiu) [reply internal/daemon/controller/handlers/scopes/scope_service_test.go:2052]: this will probably be added later, but wanted to call out we don't test completed (or running) jobs here

**Review by DanielleMiu — APPROVED** [review]

File: api/scopes/custom.go [hunk 9]
```diff
@@ -105,3 +118,43 @@ func (c *Client) RotateKeys(ctx context.Context, scopeId string, rewrapKeys bool
 	}
 	return target, nil
 }
+
+func (c *Client) ListKeyVersionDestructionJobs(ctx context.Context, scopeId string, opt ...Option) (*KeyVersionDestructionJobListResult, error) {
+	if scopeId == "" {
+		return nil, fmt.Errorf("empty scopeId value passed into ListKeyVersionDestructionJobs request")
+	}
```
[the remark this answers is not in the document]
Comment (johanbrandhorst) [reply api/scopes/custom.go:125]: I think in the CLI we consistently error when one isn't set while on the API we consistently infer global if one isn't set 😁.

File: testing/internal/e2e/scope/scope_test.go [hunk 10]
```diff
@@ -0,0 +1,231 @@
+package scope_test
```
Comment (johanbrandhorst) [inline testing/internal/e2e/scope/scope_test.go:1]: Both these tests still need to be rewritten to be independent of previous runs. Will probably run in a temporary scope instead of "global".

File: internal/db/schema/migrations/oss/postgres/56/03_create_kms_key_destructions.up.sql [hunk 11]
```diff
@@ -18,7 +18,8 @@ begin;
     key_id kms_private_id not null
       references kms_data_key_version_destruction_job (key_id)
         on delete cascade
-        on update cascade,
+        on update cascade
+        deferrable initially deferred,
```
Comment (johanbrandhorst) [inline internal/db/schema/migrations/oss/postgres/56/03_create_kms_key_destructions.up.sql:22]: This lets us try to create the job_runs before the job, which means we don't have to create the job and then delete it if there were no runs.

File: internal/db/sqltest/tests/kms/kms_data_key_destruction_job.sql [hunk 12]
```diff
@@ -26,14 +26,6 @@ begin;
       ('kdkv_______colors');
   select lives_ok('insert_new_key_id', 'insert of valid key_id in kms_data_key_version_destruction_job failed');
 
-  -- Should fail when inserting an unknown key_id
-  prepare run_insert_unknown_key_id as
-    insert into kms_data_key_version_destruction_job_run
-      (key_id, table_name, total_count)
-    values
-      ('kdkv___unknown', 'auth_token', 100);
-  select throws_ok('run_insert_unknown_key_id', '23503', null, 'insert of unknown key_id in kms_data_key_version_destruction_job_run succeeded');
```
Comment (johanbrandhorst) [inline internal/db/sqltest/tests/kms/kms_data_key_destruction_job.sql:35]: It wasn't possible to trigger this _inside_ of a transaction anymore as the consistency check is done at transaction commit time. I couldn't find an easy way to run this test at commit time so I removed it.

File: internal/kms/job/job.go [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+	const op = "kmsjob.RegisterJobs"
+
+	dataKeyVersionDestructionMonitorJob, err := newDataKeyVersionDestructionMonitorJob(ctx, kmsRepo)
+	if err != nil {
+		return errors.Wrap(ctx, err, op)
+	}
+	if err := s.RegisterJob(ctx, dataKeyVersionDestructionMonitorJob); err != nil {
+		return errors.Wrap(ctx, err, op)
+	}
+	for _, tableName := range kms.ListTablesSupportingRewrap() {
```
Comment (johanbrandhorst) [inline internal/kms/job/job.go:22]: Note: this lists the tables that are registered via `init` functions, not all tables referencing the key_id column. What guarantees that those lists are the same is a test that runs both and compares the results, which is not part of this PR.

File: testing/internal/e2e/scope/scope_test.go [hunk 14]
```diff
@@ -0,0 +1,322 @@
...
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+// TestKeyDestructionCli uses the boundary CLI to test key destruction.
+func TestKeyDestructionCli(t *testing.T) {
+	e2e.MaybeSkipTest(t)
+
+	output := boundary.AuthenticateCli()
+	require.NoError(t, output.Err, string(output.Stderr))
```
Comment (moduli) [inline testing/internal/e2e/scope/scope_test.go:22]: When you rebase with main, you would need to change this to...
```suggestion
        boundary.AuthenticateCli(t)
```

File: testing/internal/e2e/scope/scope_test.go [hunk 15]
```diff
@@ -0,0 +1,322 @@
+package scope_test
```
Comment (moduli) [inline testing/internal/e2e/scope/scope_test.go:1]: We'll have to think about file organization here. For now, I might suggest to add `scope_test.go` to the `static_test` package in `e2e/host/static` (you might not see this until you rebase with main). That way, it will automatically get run in the existing enos scenario. The alternative is to add another enos scenario, which would add some additional overhead in CI.

File: testing/internal/e2e/scope/scope_test.go [hunk 16]
```diff
@@ -0,0 +1,322 @@
...
+	var scopeReply scopes.ScopeCreateResult
+	require.NoError(t, json.Unmarshal(output.Stdout, &scopeReply))
+	t.Cleanup(func() {
+		output = e2e.RunCommand(
+			"boundary", "scopes", "delete",
+			"-id", scopeReply.Item.Id,
+			"-format", "json",
+		)
+		require.NoError(t, output.Err, string(output.Stderr))
+	})
```
Comment (moduli) [inline testing/internal/e2e/scope/scope_test.go:40]: When you rebase with main, you could change this to
```suggestion
	scopeId := boundary.CreateNewOrgCli(t)
```

File: testing/internal/e2e/scope/scope_test.go [hunk 17]
```diff
@@ -0,0 +1,322 @@
...
+	output = e2e.RunCommand(
+		"boundary", "scopes", "list-keys",
+		"-scope-id", scopeReply.Item.Id,
+		"-format", "json",
+	)
+	require.NoError(t, output.Err, string(output.Stderr))
+	var keys scopes.KeyListResult
+	err := json.Unmarshal(output.Stdout, &keys)
+	require.NoError(t, err)
+	assert.Len(t, keys.Items, 7)
```
Comment (moduli) [inline testing/internal/e2e/scope/scope_test.go:51]: What does the `7` represent here? Why do we expect 7 items?

File: testing/internal/e2e/scope/scope_test.go [hunk 18]
```diff
@@ -0,0 +1,322 @@
...
+		"-format", "json",
+	)
+	require.NoError(t, output.Err, string(output.Stderr))
+	var amReply authmethods.AuthMethodCreateResult
+	require.NoError(t, json.Unmarshal(output.Stdout, &amReply))
+	t.Cleanup(func() {
+		output = e2e.RunCommand(
+			"boundary", "auth-methods", "delete",
+			"-id", amReply.Item.Id,
+		)
```
Comment (moduli) [inline testing/internal/e2e/scope/scope_test.go:72]: nit: If you wanted to, you could change this to one line since it's shorter. I'm not sure what the preferred styling is here.
```suggestion
		output = e2e.RunCommand("boundary", "auth-methods", "delete", "-id", amReply.Item.Id)
```

File: testing/internal/e2e/scope/scope_test.go [hunk 19]
```diff
@@ -0,0 +1,322 @@
...
+		"-scope-id", scopeReply.Item.Id,
+		"-format", "json",
+	)
+	require.NoError(t, output.Err, string(output.Stderr))
+	var keys scopes.KeyListResult
+	err := json.Unmarshal(output.Stdout, &keys)
+	require.NoError(t, err)
+	assert.Len(t, keys.Items, 7)
+	for _, key := range keys.Items {
+		assert.Len(t, key.Versions, 1)
```
Comment (moduli) [inline testing/internal/e2e/scope/scope_test.go:53]: Do we care what the version value is here? Is there a default value?

File: testing/internal/e2e/scope/scope_test.go (hunk 18 above)
Comment (moduli) [reply testing/internal/e2e/scope/scope_test.go:72]: Separate question, if we're deleting the scope/org in cleanup, do we need to explicitly clean up other things in that scope/org?

File: testing/internal/e2e/scope/scope_test.go [hunk 20]
```diff
@@ -0,0 +1,322 @@
...
+		require.NoError(t, output.Err, string(output.Stderr))
+	})
+
+	output = e2e.RunCommand(
+		"boundary", "scopes", "rotate-keys",
+		"-scope-id", scopeReply.Item.Id,
+		"-format", "json",
+	)
+	require.NoError(t, output.Err, string(output.Stderr))
+	var rotate scopes.KeysRotateResult
```
Comment (moduli) [inline testing/internal/e2e/scope/scope_test.go:82]: Do we use `rotate` anywhere or is this simply just to ensure that the output from the command can be appropriately parsed/unmarshaled?

File: testing/internal/e2e/scope/scope_test.go [hunk 21]
```diff
@@ -0,0 +1,322 @@
...
+
+	ctx := context.Background()
+	sc := scopes.NewClient(client)
+
+	scope, err := sc.Create(ctx, "global", scopes.WithName("testscope"))
+	require.NoError(t, err)
+	t.Cleanup(func() {
+		_, err := sc.Delete(context.Background(), scope.Item.Id)
+		require.NoError(t, err)
+	})
```
Comment (moduli) [inline testing/internal/e2e/scope/scope_test.go:220]: When you rebase with main, this could be changed to
```suggestion
	scopeId := boundary.CreateNewOrgApi(t, ctx, client)
```

File: testing/internal/e2e/scope/scope_test.go [hunk 22]
```diff
@@ -0,0 +1,322 @@
...
+
+	keys, err := sc.ListKeys(ctx, scope.Item.Id)
+	require.NoError(t, err)
+	assert.Len(t, keys.Items, 7)
+	// Assert that all keys have the same number of versions
+	numVersions := len(keys.Items[0].Versions)
```
Comment (moduli) [inline testing/internal/e2e/scope/scope_test.go:226]: Just pointing out that this was done a little differently in the cli test (it was hard-coding numbers). I assume now that the tests are in different scopes, you can still assume there should only be 1 version here?

File: internal/kms/kms.go [hunk 23]
```diff
@@ -269,6 +273,245 @@ func (k *Kms) RotateKeys(ctx context.Context, scopeId string, opt ...Option) err
...
+			}
+			defer rows.Close()
+			var numRows uint
+			for rows.Next() {
+				if err := r.ScanRows(ctx, rows, &numRows); err != nil {
+					errors.Wrap(ctx, err, op, errors.WithMsg("failed to scan number of rows for %q", table.GetTableName()))
+				}
+			}
+			if numRows == 0 {
+				// No rows to rewrap 🎉
```
Comment (DanielleMiu) [inline internal/kms/kms.go:482]: lol love this

File: internal/kms/kms.go [hunk 24]
```diff
@@ -269,6 +273,245 @@ func (k *Kms) RotateKeys(ctx context.Context, scopeId string, opt ...Option) err
...
+			var cancel context.CancelFunc
+			ctx, cancel = context.WithTimeout(context.Background(), 5*time.Second)
+			defer cancel()
+		}
+		// Update the progress of this run
+		args := []interface{}{
+			sql.Named("key_id", run.KeyId),
+			sql.Named("table_name", tableName),
+		}
+		if _, err := k.writer.Exec(ctx, fmt.Sprintf(updateCompletedCountQueryTemplate, tableName), args); err != nil {
```
Comment (DanielleMiu) [inline internal/kms/kms.go:355]: so this query will updated the completed count to an accurate figure even if a previous run was killed and unable to update the completed count. however, this will leave an inaccurate completed count throughout the duration of this run's execution. I assume that's known + expected + and we don't really care since we only update after a run finishes anyway?

File: internal/tests/api/scopes/scope_test.go [hunk 25]
```diff
@@ -152,3 +155,102 @@ func TestErrors(t *testing.T) {
...
+	assert.Equal(t, "pending", result.State)
+
+	jobs, err = sc.ListKeyVersionDestructionJobs(ctx, "global")
+	require.NoError(t, err)
+	assert.Len(t, jobs.Items, 1)
+
+	// The configured scheduler monitoring interval is 1 second. The jobs become available 1
+	// second after the last successful run. We need to re-encrypt data in 4 different tables,
+	// and then we need to destroy the key. This job will take between 4 and 5 seconds to run,
+	// depending on the timing of the first started run.
```
Comment (DanielleMiu) [inline internal/tests/api/scopes/scope_test.go:228]: 1. only 4 tables?
2. is there any data to rewrap in tests? I didn't think that the test setup inserted any default data into the tables, and I didn't see you insert any in the test itself

File: internal/tests/api/scopes/scope_test.go [hunk 26]
```diff
@@ -152,3 +155,102 @@ func TestErrors(t *testing.T) {
...
+		}
+	}
+
+	result, err = sc.DestroyKeyVersion(ctx, "global", destroyKeyVersion.Id)
+	require.NoError(t, err)
+	assert.Equal(t, "pending", result.State)
+
+	jobs, err = sc.ListKeyVersionDestructionJobs(ctx, "global")
+	require.NoError(t, err)
+	assert.Len(t, jobs.Items, 1)
```
Comment (DanielleMiu) [inline internal/tests/api/scopes/scope_test.go:223]: if you didn't insert any data for rewrapping, wouldn't destroy return a completed status? per [this logic from destroy key version](https://github.com/hashicorp/boundary/pull/2498/files#diff-65fa95abdd78accdc6b1f3ac98751fe4700c8abc1ee696b8d6ae7e5403af0831R511-R517)?

File: testing/internal/e2e/scope/scope_test.go (hunk 17 above)
Comment (johanbrandhorst) [reply testing/internal/e2e/scope/scope_test.go:51]: We have 7 different key purposes in each scope, so we get 7 keys per scope.

File: testing/internal/e2e/scope/scope_test.go (hunk 18 above)
Comment (johanbrandhorst) [reply testing/internal/e2e/scope/scope_test.go:72]: I like to be explicit about it, even though it probably doesn't need to be done 😄

File: testing/internal/e2e/scope/scope_test.go (hunk 19 above)
Comment (johanbrandhorst) [reply testing/internal/e2e/scope/scope_test.go:53]: The version is a struct which has ID, CreateTime on it, so it's not easy to compare against something.

**Review by DanielleMiu — CHANGES_REQUESTED** [review]

File: testing/internal/e2e/scope/scope_test.go (hunk 20 above)
Comment (johanbrandhorst) [reply testing/internal/e2e/scope/scope_test.go:82]: Hm, you're right, we can probably just remove this

File: testing/internal/e2e/scope/scope_test.go (hunk 22 above)
Comment (johanbrandhorst) [reply testing/internal/e2e/scope/scope_test.go:226]: Good shout, I missed changing this back

File: internal/daemon/controller/handlers/scopes/scope_service.go [hunk 27]
```diff
@@ -432,6 +434,76 @@ func (s Service) RotateKeys(ctx context.Context, req *pbs.RotateKeysRequest) (*p
...
+	if err != nil {
+		if errors.Match(errors.T(errors.KeyNotFound), err) {
+			return nil, handlers.NotFoundErrorf("unknown key_version_id %q", req.KeyVersionId)
+		}
+		return nil, err
+	}
+	state := "completed"
+	if !destroyed {
+		state = "pending"
+	}
```
[the remark this answers is not in the document]
Comment (johanbrandhorst) [reply internal/daemon/controller/handlers/scopes/scope_service.go:501]: I don't think there's much reason to introduce an enum for something as simple as this. It could be a boolean, but this was the API definition we agreed on, which I think is nicer than a boolean and certainly more flexible.

File: internal/tests/api/scopes/scope_test.go (hunk 25 above)
Comment (johanbrandhorst) [reply internal/tests/api/scopes/scope_test.go:228]: Yep! There's 8 pieces of encrypted data created by the test controller setup (creating some tokens, users, auth methods, etc) spread across 4 tables in the global scope.

File: internal/tests/api/scopes/scope_test.go (hunk 26 above)
Comment (johanbrandhorst) [reply internal/tests/api/scopes/scope_test.go:223]: It would, yep! But since there's default data it doesn't.

File: internal/kms/kms.go (hunk 24 above)
Comment (johanbrandhorst) [reply internal/kms/kms.go:355]: Indeed, we don't do any mid-run updates, only at the end of the run. Since we break up the rewrapping per-table I think that's still okay as the user will get updates every time a table finishes. We could implement mid-run updates in a future update.

File: internal/kms/kms.go [hunk 28]
```diff
@@ -269,6 +273,245 @@ func (k *Kms) RotateKeys(ctx context.Context, scopeId string, opt ...Option) err
...
+func (k *Kms) ListDataKeyVersionDestructionJobs(ctx context.Context, scopeId string, _ ...Option) ([]*DataKeyVersionDestructionJobProgress, error) {
+	var jobs []*DataKeyVersionDestructionJobProgress
+	if err := k.reader.SearchWhere(ctx, &jobs, "scope_id=?", []interface{}{scopeId}); err != nil {
+		return nil, err
+	}
+	return jobs, nil
+}
+
+// MonitorTableRewrappingRuns checks for pending rewrapping runs for the
+// specified table name, and attempts to become the running run and
```
[the remark this answers is not in the document]
Comment (johanbrandhorst) [reply internal/kms/kms.go:287]: The terminology is a bit of a struggle 😅. I'll update it.

File: internal/kms/kms.go [hunk 29]
```diff
@@ -269,6 +273,245 @@ func (k *Kms) RotateKeys(ctx context.Context, scopeId string, opt ...Option) err
...
+	run := allocDataKeyVersionDestructionJobRun()
+	err := k.reader.LookupWhere(ctx, &run, "is_running=true and table_name!=?", []interface{}{tableName})
+	switch {
+	case err == nil:
+		// Another run was already running
+		return nil
+	case errors.Match(errors.T(errors.RecordNotFound), err):
+		// Great, no running run, lets continue
+	default:
+		return errors.Wrap(ctx, err, op, errors.WithMsg("failed to find any running runs"))
```
[the remark this answers is not in the document]
Comment (johanbrandhorst) [reply internal/kms/kms.go:313]: Nah, this is for an unexpected error like no database connection or something like that.

File: testing/internal/e2e/scope/scope_test.go (hunk 15 above)
Comment (johanbrandhorst) [reply testing/internal/e2e/scope/scope_test.go:1]: I'll move it

File: internal/db/sqltest/tests/kms/kms_data_key_destruction_job.sql (hunk 12 above)
Comment (johanbrandhorst) [reply internal/db/sqltest/tests/kms/kms_data_key_destruction_job.sql:35]: No, since it's a direct database schema test, we don't really have another place to test this, and frankly I'm not sure it's necessary.

File: internal/db/schema/migrations/oss/postgres/56/03_create_kms_key_destructions.up.sql (hunk 11 above)
Comment (johanbrandhorst) [reply internal/db/schema/migrations/oss/postgres/56/03_create_kms_key_destructions.up.sql:22]: It happens in the same transaction as the job runs in `DestroyKeyVersion`: https://github.com/hashicorp/boundary/pull/2498/files#diff-65fa95abdd78accdc6b1f3ac98751fe4700c8abc1ee696b8d6ae7e5403af0831R504. What do you mean that they will be deleted? It will error at commit and rollback if these constraints are not met, it won't delete anything silently.

File: internal/daemon/controller/handlers/scopes/scope_service.go (hunk 27 above)
[the remark this answers is not in the document]
Comment (johanbrandhorst) [reply internal/daemon/controller/handlers/scopes/scope_service.go:501]: I assume by enum you mean something like
```go
type KeyDestructionState int

const (
	KeyDestructionStateUnspecified KeyDestructionState = iota
	KeyDestructionStatePending
	KeyDestructionStateCompleted
)

func (k KeyDestructionState) String() string {
	switch k {
	case KeyDestructionStatePending:
		return "pending"
	case KeyDestructionStateCompleted:
		return "completed"
	default:
		return "unspecified"
	}
}
```

It's quite a bit of overhead for little gain here, I think, and it actually introduces the issue of using the zero value of the type by mistake.

File: internal/kms/kms.go [hunk 30]
```diff
@@ -269,6 +273,254 @@ func (k *Kms) RotateKeys(ctx context.Context, scopeId string, opt ...Option) err
...
+		for _, table := range tables {
+			rows, err := r.Query(ctx, fmt.Sprintf(findAffectedRowsForKeyQueryTemplate, table.GetTableName()), []interface{}{keyVersionId})
+			if err != nil {
+				errors.Wrap(ctx, err, op, errors.WithMsg("failed to get affected rows for %q", table.GetTableName()))
+			}
+			defer rows.Close()
+			var numRows uint
+			for rows.Next() {
+				if err := r.ScanRows(ctx, rows, &numRows); err != nil {
+					errors.Wrap(ctx, err, op, errors.WithMsg("failed to scan number of rows for %q", table.GetTableName()))
```
Comment (DanielleMiu) [inline internal/kms/kms.go:480]: these two errors are not returned?

Comment (johanbrandhorst) [reply internal/kms/kms.go:480]: Hm, oops!
